File: leet/0086-partition-list/partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	dummy := ListNode{Next: head}
	
	before := &dummy
	split := head
	
	for split != nil && split.Val < x {
		before = split
		split = split.Next
	}
	
    if(split == nil || split.Next == nil) {
        return dummy.Next
    }
    
	last := split
	current := split.Next
	
	for current != nil {
		if(current.Val < x){
			last.Next = current.Next
			before.Next = current
			current.Next = split
			
			before = before.Next
			current = last.Next
        }else{
            last = current
            current = current.Next
        }
	}
	
	return dummy.Next
}

File: leet/0086-partition-list/partition-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
    if (!head) return head
    
    let firstLess
    let less
    let firstGreat
    let great
    
    while(head){
        if (head.val < x){
            if (!less){
                less = head
                firstLess = head
            }else{
                less.next = head
                less = less.next
            }
        }else{
            if(!great){
                great = head
                firstGreat = head
            }else{
                great.next = head
                great = great.next
            }
        }        
        
        head = head.next
    }
    
    if(less) less.next = firstGreat
    if(great) great.next = null
    return firstLess || firstGreat
};
